bread/layout/components/datatable.py

```diff
@@ -57,19 +60,27 @@ def extractsortinglink(context, element):
     return aslink_attributes(hg.F(extractsortinglink))
 
 
+@dataclass
+class DataTableColumn:
+    header: Any
+    cell: Any
+    sortingname: str = None
+    enable_row_click: bool = True
+
+
 class DataTable(hg.BaseElement):
```

Comment:
Since we deal with immutable data here NamedTuple might be preferable here. What do you think?

Reply:
Also, see my comment on https://github.com/basxsoftwareassociation/basxconnect/pull/68#pullrequestreview-658555347 regarding an additional helper-method to prevent the use of pure strings. It might simplify some of the data-tables code. I copy-paste it here:

> Looks good to me, much more clear with the DataTableColumn. If we already use DataTableColumn, should we maybe enforce it and prohibit the use of plain strings? A static-method-shortcut like DataTableColumn.from_modelfield(fieldname) would be interesting I think.

And we could also check the columns argument in ```DataTable.__init__``` to be of the correct type (and then people say "duuhhh, that is what statically typed languages are for" :D and I would agree). There is already some type checking happening in the DataTables code, to catch misconfigured datatables early.

Reply:
I agree that no one should mutate an instance of this class ;) . I also thought of the named tuple, but when I saw the syntax it looked a lot like the advanced reflection-like features of python that I prefer not using. (Declaring the name of a field using a string...). Maybe I should add the `frozen=True` annotation. But then, it feels to go against the Python mindset to _enforce_ immutability. I think in Python people need to act responsibly, otherwise we are heading to disaster anyway... 

I like the idea of a static `DataTableColumn.from_modelfield(fieldname, model)` method!

To be honest, I am against these type checks in Python because type hints together with static analysis detect this kind of errors without adding boilerplate code for type checking. If we _have_ to enforce types, I'd much rather use a decorator like this one: https://stackoverflow.com/a/32685519 . What I love about Python is that there is almost no boilerplate, which potentially makes code very readable, so I'd rather not give that up.

Reply:
The immutability thing is not super important I think, and yes, python programmers are not supposed to behave too stupid :D. In any case we could use ```typing.NamedTuple``` this is not using strings (at least not directly, when running the code it is anyways because the types are in all cases constructed at runtime) and we could also use type hints with that.

Yes, the check of string vs DataTableColumn is not necessairily important, so I am fine without that anyways.

However, I should add: The problem with the "only type hints and static analysis" approach (or any type system for that matter) is that it is never expressiv enough to do all the checks for you. I think it is comparable to statically typed languages to some degree: Before "Nullable" became a thing in C# and Java we wrote a lot of "dynamic type checking" with ```if(arg1 == null) raise IllegalParameterException("Arg1 cannot be null")``` and similar.
Now, another problem with type hints in Python is that type hints are (and will always be, as far as I understand the propsal) only comment to the interpreter and do not influence the behaviour of the program in any way. This means type hints only work for us in our projects if you have mypy or similar integrated in the development workflow. Any other programer using this code (and this is crucial for framework and library code) will need to debug things if he doesn't run a type-checker first.
Therefore, I suggest that dynamic integrity checks are a very helpfull thing and we should keep other users of our APIs in mind.
Maybe another approach to look at it, more generally, would be: It is often helpfull to ensure certain invariants are true in order to make the rest of the code simpler because we can always rely on these invariants being true.

Reply:
Oh great, I didn't know about `typing.NamedTuple`! I like the syntax of that one :)
I agree that static analysis doesn't catch everything and that for stuff it doesn't catch, dynamic checks are very useful.
I just assumed (and hoped) Python is moving to a direction where most people will notice if their usage of a function and the type hints are not consistent. If this is not the case (and you know that better than me), I am happy to check the types manually.

Reply:
Right! So, as I just mentioned in the ```htmlgenerator``` PR: Type hinting is still very unadapted. I think we can profit from from more type hints. But, I would still keep it a bit on the "exploring" side, also for this project. In Python it is often good to follow the mainstream a bit, be it for tooling, readability or just compatability (the black-formatter was an interesting example of seeing that kind of "adoption" process throughout the wider Python community).

The ```htmlgenerator``` project is a good place to explore because it is a very standalone and small code base.
I think if we want to do more with type hinting inside bread we should integrate it into the GitHub CI and maybe start with smaller packages or modules which are more standalone. PM me if you want add more of it in existing code, I can give you a hint where to start.